modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotLocalFileSystemStorage.scala

```diff
@@ -18,24 +18,31 @@ import org.tessellation.storage.LocalFileSystemStorage
 import better.files.File
 import fs2.io.file.Path
 import io.estatico.newtype.ops._
+import org.typelevel.log4cats.slf4j.Slf4jLogger
 
 final class SnapshotLocalFileSystemStorage[F[_]: Async: KryoSerializer, S <: Snapshot] private (path: Path)
     extends LocalFileSystemStorage[F, Signed[S]](path) {
 
+  private val logger = Slf4jLogger.getLogger[F]
+
   def write(snapshot: Signed[S]): F[Unit] = {
     val ordinalName = toOrdinalName(snapshot.value)
 
     toHashName(snapshot.value).flatMap { hashName =>
       (exists(ordinalName), exists(hashName)).mapN {
-        case (ordinalExists, hashExists) =>
-          if (ordinalExists || hashExists) {
-            (new Throwable("Snapshot already exists under ordinal or hash filename")).raiseError[F, Unit]
-          } else {
-            write(hashName, snapshot) >> link(hashName, ordinalName)
-          }
-      }.flatten
-    }
-
+        case (true, true) =>
+          new Exception(s"Snapshot exists for $ordinalName and $hashName").raiseError[F, Unit]
+        case (true, false) =>
+          new Exception(s"Snapshot ordinal file $ordinalName exists but hash missing").raiseError[F, Unit]
+
+        case (false, true) =>
+          logger.warn(s"Snapshot hash file $hashName exists but ordinal missing; linking to $ordinalName") >>
+            link(hashName, ordinalName)
+
+        case (false, false) =>
+          write(hashName, snapshot) >> link(hashName, ordinalName)
+      }
```

Comment:
I agree with the improved logic for logging. I would try though to log in a single message with ordinal, shortened hash name and two conditions s"savedAsOrdinal=$ordinalExists savedAsHash=$hashExists".

Regarding the fixing logic I think it will maybe fix a very particular small case where we successfully save under hash but then for some reason fail to link. If we have corrupted state under hash we will link but the snapshot under hash will not be fixed. Also we are checking if the snapshot is saved under ordinal and hash without locking so theoretically a state could change - still I think linking in such situation should simply fail if a file already exists, so maybe that's not a problem. Overwriting snapshot by hash in such a situation as well could be a better improvement. But then even better solution would be to handle the cleanup after failed writing attempt (I think such cleanup is not happening automatically but best to check) to try to not end in such a state at all, we would also need to synchronously lock both files I think. Even better would be to only save snapshots on disk by hash and have some transactional storage on disk for the ordinal -> hash mapping (overwriting snapshot by hash is safe, overwriting by ordinal isn't hence the need for transactional update for ordinal).

So in general I'm ambivalent regarding that change as I think it will probably not have a substantial impact and won't totally solve any of the issues we have.

Reply:
> I agree with the improved logic for logging. I would try though to log in a single message with ordinal, shortened hash name and two conditions s"savedAsOrdinal=$ordinalExists savedAsHash=$hashExists".
> 
> Regarding the fixing logic I think it will maybe fix a very particular small case where we successfully save under hash but then for some reason fail to link. If we have corrupted state under hash we will link but the snapshot under hash will not be fixed. Also we are checking if the snapshot is saved under ordinal and hash without locking so theoretically a state could change - still I think linking in such situation should simply fail if a file already exists, so maybe that's not a problem. Overwriting snapshot by hash in such a situation as well could be a better improvement. But then even better solution would be to handle the cleanup after failed writing attempt (I think such cleanup is not happening automatically but best to check) to try to not end in such a state at all, we would also need to synchronously lock both files I think. Even better would be to only save snapshots on disk by hash and have some transactional storage on disk for the ordinal -> hash mapping (overwriting snapshot by hash is safe, overwriting by ordinal isn't hence the need for transactional update for ordinal).
> 
> So in general I'm ambivalent regarding that change as I think it will probably not have a substantial impact and won't totally solve any of the issues we have.

Sorry if this is completely incorrect with what you are trying to accomplish... I think I understand what you guys are talking about... but it's possible I'm missing the entire context.  But here is an idea that maybe could help...

Could it possibly be helpful to rename the file to something like "{ordinal}-{hash}" after the snapshot has completed verification and linking. I could see this being useful in the fact that you would only have a properly named file once it has completely been verified and linked. If it hasn't been verified it would only be named "{ordinal}" without a dash and no hash, which would instantly indicate a file that was not fully validated.